### ISSUE TITLE
Tensor size for `masked_fill` exceeds the limit supported by the MPS backend: must be less than 2**32 elements

### ISSUE THREAD

**Issue by rusnov** [pytorch/pytorch#143477]
### 🐛 Describe the bug

I get the following error, when using `masked_fill` on larger tensors. See error and the minimal code below.

**Error:**
```
/AppleInternal/Library/BuildRoots/.../Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:850: failed assertion `[MPSNDArray initWithDevice:descriptor:isTextureBacked:] Error: total bytes of NDArray > 2**32'
```

**Code:**
```python
import torch
device = torch.device("mps")
mask_bool = torch.triu(torch.ones(1024, 1024, device=device), diagonal=1).bool()
attn_scores = torch.rand(48, 25, 1024, 1024, device=device)
attn_scores.masked_fill_(mask_bool, 0)
```

### Versions

```
PyTorch version: 2.6.0.dev20241217
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.2 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.6)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.10 | packaged by conda-forge | (main, Oct 16 2024, 01:26:25) [Clang 17.0.6 ] (64-bit runtime)
Python platform: macOS-15.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M4 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.13.1
[pip3] torch==2.6.0.dev20241217
[pip3] torchaudio==2.6.0.dev20241217
[pip3] torchvision==0.22.0.dev20241217
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] optree                    0.13.1                   pypi_0    pypi
[conda] torch                     2.6.0.dev20241217          pypi_0    pypi
[conda] torchaudio                2.6.0.dev20241217          pypi_0    pypi
[conda] torchvision               0.22.0.dev20241217          pypi_0    pypi
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by BoBo0037** [comment]
please fix this issue ...

**Comment by jhavukainen** [comment]
@rusnov Thanks for reporting the issue and the repro case. Can confirm this is due to the current 32-bit indexing limit in MPS and we need to circumvent it by tiling the masking operation to work with the underlying MTLBuffer in segments in order to not exceed the indexing limit.

I'll work on a fix for this.

**Comment by malfet** [comment]
I wonder if better solution (both performance and readability-wise, would be to re-implement masked_fill as Metal shader)

**Comment by malfet** [comment]
And with https://github.com/pytorch/pytorch/pull/143893 applied I can compile it away
```python
import torch

def foo(x, y):
   x.masked_fill_(y, 0)

device = torch.device("mps")
mask_bool = torch.triu(torch.ones(1024, 1024, device=device), diagonal=1).bool()
attn_scores = torch.rand(48, 25, 1024, 1024, device=device)

torch.compile(foo)(attn_scores, mask_bool)
```

**Comment by fabiopolimeni** [comment]
Sorry if this may not be the same issue (but possibly yes), but I am getting the same crash 

```
com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:850: failed assertion `[MPSNDArray initWithDevice:descriptor:isTextureBacked:] Error: total bytes of NDArray > 2**32'
```

with the following stacktrace:

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               Python [14811]
Path:                  /Library/Frameworks/Python.framework/Versions/3.11/Resources/Python.app/Contents/MacOS/Python
Identifier:            org.python.python
Version:               3.11.6 (3.11.6)
Code Type:             ARM-64 (Native)
Parent Process:        zsh [4967]
Responsible:           Hyper [4960]
User ID:               501

Date/Time:             2025-01-06 05:15:38.4653 +0100
OS Version:            macOS 15.1.1 (24B91)
Report Version:        12
Anonymous UUID:        0EF2994A-C5DF-BBC0-5E24-9A4628DEDB38

Sleep/Wake UUID:       233E8275-D4D3-4AB6-896D-CC636A906E78

Time Awake Since Boot: 41000 seconds
Time Since Wake:       5438 seconds

System Integrity Protection: enabled

Crashed Thread:        1

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 6 Abort trap: 6
Terminating Process:   Python [14811]

Thread 0::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x19619befc kevent + 8
1   select.cpython-311-darwin.so  	       0x10536cb90 select_kqueue_control_impl + 592
2   Python                        	       0x105bca948 _PyEval_EvalFrameDefault + 63264
3   Python                        	       0x105bb9f7c PyEval_EvalCode + 276
4   Python                        	       0x105c3339c pyrun_file + 240
5   Python                        	       0x105c32b6c _PyRun_SimpleFileObject + 288
6   Python                        	       0x105c320ec _PyRun_AnyFileObject + 232
7   Python                        	       0x105c5b92c pymain_run_file_obj + 220
8   Python                        	       0x105c5b110 pymain_run_file + 72
9   Python                        	       0x105c5a924 Py_RunMain + 1828
10  Python                        	       0x105c5baa8 pymain_main + 52
11  Python                        	       0x105c5c2e0 Py_BytesMain + 40
12  dyld                          	       0x195e54274 start + 2840

Thread 1 Crashed:
0   libsystem_kernel.dylib        	       0x19619e600 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x1961d6f70 pthread_kill + 288
2   libsystem_c.dylib             	       0x1960e3908 abort + 128
3   libsystem_c.dylib             	       0x1960e2c1c __assert_rtn + 284
4   Metal                         	       0x1a1139870 MTLReportFailure.cold.1 + 48
5   Metal                         	       0x1a1115198 MTLReportFailure + 448
6   MPSCore                       	       0x1a2116830 0x1a20be000 + 362544
7   MPSCore                       	       0x1a2116c94 0x1a20be000 + 363668
8   MetalPerformanceShadersGraph  	       0x22ffbd5c4 -[MPSGraphTensorData initWithMTLBuffer:shape:dataType:] + 480
9   libtorch_cpu.dylib            	       0x16ff1b620 at::native::mps::Placeholder::Placeholder(MPSGraphTensor*, at::Tensor const&, NSArray<NSNumber*>*, bool, MPSDataType, bool) + 2476
10  libtorch_cpu.dylib            	       0x17000e138 at::native::structured_softmax_mps_out::impl(at::Tensor const&, long long, bool, at::Tensor const&) + 1612
11  libtorch_cpu.dylib            	       0x16cfb72a8 at::(anonymous namespace)::wrapper_MPS__softmax(at::Tensor const&, long long, bool) + 116
12  libtorch_cpu.dylib            	       0x16c26e778 at::_ops::_softmax::call(at::Tensor const&, long long, bool) + 292
13  libtorch_cpu.dylib            	       0x16b86a418 at::native::softmax(at::Tensor const&, long long, std::__1::optional<c10::ScalarType>) + 264
14  libtorch_cpu.dylib            	       0x16c26d8ec at::_ops::softmax_int::call(at::Tensor const&, long long, std::__1::optional<c10::ScalarType>) + 296
15  libtorch_cpu.dylib            	       0x16bcdc334 at::native::_safe_softmax(at::Tensor const&, long long, std::__1::optional<c10::ScalarType>) + 60
16  libtorch_cpu.dylib            	       0x16c30189c at::_ops::_safe_softmax::call(at::Tensor const&, long long, std::__1::optional<c10::ScalarType>) + 296
17  libtorch_cpu.dylib            	       0x16bce1ecc at::native::_scaled_dot_product_attention_math(at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::optional<at::Tensor> const&, double, bool, std::__1::optional<at::Tensor> const&, std::__1::optional<double>, bool) + 6772
18  libtorch_cpu.dylib            	       0x16c414840 at::_ops::_scaled_dot_product_attention_math::call(at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::optional<at::Tensor> const&, double, bool, std::__1::optional<at::Tensor> const&, std::__1::optional<double>, bool) + 416
19  libtorch_cpu.dylib            	       0x16bcdd390 at::native::scaled_dot_product_attention(at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::optional<at::Tensor> const&, double, bool, std::__1::optional<double>, bool) + 2856
20  libtorch_cpu.dylib            	       0x16c1aadbc at::_ops::scaled_dot_product_attention::call(at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::optional<at::Tensor> const&, double, bool, std::__1::optional<double>, bool) + 388
21  libtorch_python.dylib         	       0x109a4f408 torch::autograd::THPVariable_scaled_dot_product_attention(_object*, _object*, _object*) + 832
```

Is this related? This happens somewhere when trying do some inference on a LTX-Video pipeline on a Mac M4 Max 48GB of overall memory.

**Comment by ZainRizvi** [comment]
Seems like this was already fixed before the branch cut. It succeeded on `torch==2.7.0.dev20250312` as well as the latest rc

```
(release2.7) ~/test/release2.7/.venv/lib/python3.12/site-packages/torch/lib python
Python 3.12.5 (main, Aug 14 2024, 04:32:18) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
/Users/zainr/test/release2.7/.venv/lib/python3.12/site-packages/torch/_subclasses/functional_tensor.py:276: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/utils/tensor_numpy.cpp:81.)
  cpu = _conversion_method_template(device=torch.device("cpu"))
>>> device = torch.device("mps")
>>> mask_bool = torch.triu(torch.ones(1024, 1024, device=device), diagonal=1).bool()
>>> attn_scores = torch.rand(48, 25, 1024, 1024, device=device)
>>> attn_scores.masked_fill_(mask_bool, 0)
tensor([[[[0.4690, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.3107, 0.8356, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.0364, 0.0291, 0.0664,  ..., 0.0000, 0.0000, 0.0000],
          ...,
          [0.0581, 0.5510, 0.5407,  ..., 0.0711, 0.0000, 0.0000],
          [0.0695, 0.5823, 0.1001,  ..., 0.1280, 0.9604, 0.0000],
          [0.4948, 0.1125, 0.4361,  ..., 0.4127, 0.1208, 0.8033]],

         [[0.6927, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.7147, 0.4760, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.5352, 0.6932, 0.6671,  ..., 0.0000, 0.0000, 0.0000],
          ...,
          [0.1711, 0.2745, 0.0447,  ..., 0.2753, 0.0000, 0.0000],
          [0.6427, 0.4958, 0.4723,  ..., 0.4911, 0.0919, 0.0000],
          [0.7332, 0.1662, 0.7073,  ..., 0.3584, 0.1191, 0.6283]],

         [[0.1825, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.3658, 0.3885, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
          [0.7345, 0.9455, 0.7635,  ..., 0.0000, 0.0000, 0.0000],
          ...,
          [0.9746, 0.3109, 0.7781,  ..., 0.9730, 0.0000, 0.0000],
          [0.6107, 0.2254, 0.0466,  ..., 0.5542, 0.3392, 0.0000],
          [0.5938, 0.6827, 0.1850,  ..., 0.3648, 0.2013, 0.2783]],

         ...,
```

**Comment by theobouwman** [comment]
@ZainRizvi great! I do notice that `pin_memory` is not available anymore in the `torch==2.7.0.dev20250312` version?
```
UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, then device 
pinned memory won't be used.
```

**Comment by ZainRizvi** [comment]
pin_memory [seems to](https://github.com/pytorch/pytorch/issues/150179) only be supported for CPU.  @albanD can confirm
